include/swift/SIL/SILInstruction.h

```diff
@@ -5625,39 +5625,62 @@ class TupleElementAddrInst
   }
 };
 
-/// Extract a physical, fragile field out of a value of struct type.
-class StructExtractInst
-  : public UnaryInstructionBase<SILInstructionKind::StructExtractInst,
-                                SingleValueInstruction>
-{
-  friend SILBuilder;
-
-  VarDecl *Field;
+/// A common base for instructions that require a cached field index.
+///
+/// The field's ordinal value is used by the Projection abstraction, which is
+/// the basis of efficiently comparing and sorting access paths in SIL. It must
+/// be cached before it is used by any SIL passes, including the verifier, or it
+/// will cause widespread quadratic complexity. The type system does not (yet)
+/// cache a VarDecl's field index because it is unstable until the type is
+/// finalized.
```

Comment:
This comment has a lot of implicit knowledge about SIL/SILOptimizer how it is used in it. I think we should be more explicit. My thoughts:

1. I would make it clear that the intent is for this to be used with "projection-like" instructions in the \brief.
2. I would make the 2nd sentence of the descriptive have more context especially around the type system comment.

My thought:

```
A common base for Projection-compatible instructions that use a cached field index to avoid quadratic complexity.

The Projection abstraction uses a compatible instruction's ordinal value to efficiently compare and sort access paths in SIL. Without using a cache, we must iterate over parts of the AST to find such an index resulting in widespread quadratic complexity.

FIXME: The reason why this is done is that the Decl TypeChecker does not currently cache a VarDecl's field index due to a VarDecl's field index being unstable until the type is finalized. If/when this is changed, we can eliminate this class and just access the pre-computed index on the actual VarDecl.

Reply:
Thanks. The wording I chose below is deliberate. If it doesn't make sense to you let me know.
```
/// A common base for instructions that require a cached field index.
///
/// "Field" is a term used here to refer to the ordered, accessible stored
/// properties of a class or struct.
///
/// The field's ordinal value is the basis of efficiently comparing and sorting
/// access paths in SIL. For example, whenever a Projection object is created,
/// is stores the field index. Finding the field index initially requires
/// searching the type declaration's array of all stored properties. If this
/// index is not cached, it will cause widespread quadratic complexity in any
/// pass that queries projections, including the SIL verifier.
///
/// FIXME: This cache may not be necessary if the Decl TypeChecker instead
/// caches a field index in the VarDecl itself. This solution would be superior
/// because it would allow constant time lookup of either the VarDecl or the
/// index from a single pointer without referring back to a projection
/// instruction.
```

Reply:
Amazing! +1! Only one typo:

```
For example, whenever a Projection object is created, is stores the field index.
```

=>

```
For example, whenever a Projection object is created, it stores the field index.
```